File: Binary Search/Objective-C/BinarySearch.m
//
//  BinarySearch.m
//  Algorithms
//
//  Copyright © 2017 Grigory Lutkov <Friend.LGA@gmail.com>
//

#import "BinarySearch.h"

@implementation BinarySearch

+ (NSUInteger)indexOfNumber:(NSNumber *)number inSortedArray:(NSArray<NSNumber *> *)array inPlace:(BOOL)inPlace {
    if (number == nil || array == nil || array.count == 0) {
        return NSNotFound;
    }

    if (inPlace) {
        return [self inPlaceIndexOfNumber:number inSortedArray:array range:NSMakeRange(0, array.count)];
    }
    else {
        return [self notInPlaceIndexOfNumber:number inSortedArray:array indexOffset:0];
    }
}

#pragma mark - In Place

+ (NSUInteger)inPlaceIndexOfNumber:(NSNumber *)number inSortedArray:(NSArray<NSNumber *> *)array range:(NSRange)range {
    if (range.length == 0) {
        return NSNotFound;
    }

    NSUInteger rangeLocation = range.location;
    NSUInteger rangeMiddleIndex = range.length / 2;
    NSUInteger middleIndex = rangeLocation + rangeMiddleIndex;
    NSNumber *middleNumber = array[middleIndex];

    switch ([number compare:middleNumber]) {
        case NSOrderedSame: {
            return middleIndex;
        }
        case NSOrderedAscending: {
            return [self inPlaceIndexOfNumber:number inSortedArray:array range:NSMakeRange(rangeLocation, rangeMiddleIndex)];
        }
        case NSOrderedDescending: {
            NSUInteger nextIndex = middleIndex + 1;
            return [self inPlaceIndexOfNumber:number inSortedArray:array range:NSMakeRange(nextIndex, array.count - nextIndex)];
        }
    }
}

#pragma mark - Not In Place

+ (NSUInteger)notInPlaceIndexOfNumber:(NSNumber *)number inSortedArray:(NSArray<NSNumber *> *)array indexOffset:(NSUInteger)indexOffset {
    NSUInteger arrayCount = array.count;

    if (arrayCount == 0) {
        return NSNotFound;
    }

    NSUInteger middleIndex = arrayCount / 2;
    NSNumber *middleNumber = array[middleIndex];

    switch ([number compare:middleNumber]) {
        case NSOrderedSame: {
            return middleIndex + indexOffset;
        }
        case NSOrderedAscending: {
            return [self notInPlaceIndexOfNumber:number inSortedArray:[array subarrayWithRange:NSMakeRange(0, middleIndex)] indexOffset:indexOffset];
        }
        case NSOrderedDescending: {
            NSUInteger nextIndex = middleIndex + 1;
            return [self notInPlaceIndexOfNumber:number inSortedArray:[array subarrayWithRange:NSMakeRange(nextIndex, arrayCount - nextIndex)] indexOffset:indexOffset + nextIndex];
        }
    }
}

@end

File: Binary Search/Swift/BinarySearch.swift
//
//  BinarySearch.swift
//  Algorithms
//
//  Copyright © 2017 Grigory Lutkov <Friend.LGA@gmail.com>
//

import Foundation

struct BinarySearch {

    static func index<T: Comparable>(of number: T?, inSortedArray array: [T?]?, inPlace: Bool) -> UInt? {
        if number == nil || array == nil || array?.count == 0 {
            return nil
        }

        if inPlace {
            return inPlaceIndex(of: number!, inSortedArray: array!, range: 0..<UInt(array!.count))
        }
        else {
            return notInPlaceIndex(of: number!, inSortedArray: array!, indexOffset: 0)
        }
    }

    // MARK: - In Place

    static private func inPlaceIndex<T: Comparable>(of number: T, inSortedArray array: [T?], range: Range<UInt>) -> UInt? {
        if range.count == 0 {
            return nil
        }

        let rangeUpperBound = range.upperBound
        let middleIndex = (range.lowerBound + rangeUpperBound) / 2
        let middleNumber = array[Int(middleIndex)]

        if middleNumber == nil {
            return nil
        }

        if number == middleNumber {
            return middleIndex
        }
        else if number < middleNumber! {
            return inPlaceIndex(of: number, inSortedArray: array, range: 0..<middleIndex)
        }
        else {
            return inPlaceIndex(of: number, inSortedArray: array, range: (middleIndex + 1)..<rangeUpperBound)
        }
    }

    // MARK: - Not In Place

    static private func notInPlaceIndex<T: Comparable>(of number: T, inSortedArray array: [T?], indexOffset: UInt) -> UInt? {
        let arrayCount = array.count

        if arrayCount == 0 {
            return nil
        }

        let middleIndex = arrayCount / 2
        let middleNumber = array[middleIndex]

        if middleNumber == nil {
            return nil
        }

        if number == middleNumber {
            return UInt(middleIndex) + indexOffset
        }
        else if number < middleNumber! {
            return notInPlaceIndex(of: number, inSortedArray: Array(array[0..<middleIndex]), indexOffset: indexOffset)
        }
        else {
            let nextIndex = middleIndex + 1
            return notInPlaceIndex(of: number, inSortedArray: Array(array[nextIndex..<arrayCount]), indexOffset: indexOffset + UInt(nextIndex))
        }
    }

}
